LOG-1344: Multi clusterlogforwarder proposal (#1410)
openshift/enhancements: enhancements/cluster-logging/multi-cluster-log-forwarder.md

---
title: multi-cluster-log-forwarder
authors:
  - @jcantril
reviewers:
  - @alanconway, Red Hat Logging Architect
approvers: 
  - @alanconway
api-approvers:
  - @alanconway 
creation-date: 2023-02-23
last-updated: 2023-02-23
tracking-link: 
  - https://issues.redhat.com/browse/LOG-1344
see-also:
  - 
replaces:
  - 
---

# Multi ClusterLogForwarder
## Summary

Log forwarding is functionally a "cluster singleton" where the operator explicitly only reconcilies a **ClusterLogForwarder** in the  namespace *openshift-logging* named *instance*.  This enhancement removes that restriction to allow administrators to define multiple instance of **ClusterLogForwarder** while retaining the legacy behavior.

## Motivation

> This section is for explicitly listing the motivation, goals and non-goals of
> this proposal. Describe why the change is important and the benefits to users.

Reviewer: ```suggestion
```

### User Stories

* As an administrator of a Red Hat managed cluster, I want to RBAC my log forwarder configuration from customer admins so they can take ownership of their log forwarder needs without being able to modify mine.

> * As an administrator of HyperShift, I want to deploy individual log forwarders to isolate audit log collection of each managed control plane.

Reviewer: ```suggestion
* As an administrator of Hosted Control Planes, I want to deploy individual log forwarders to isolate audit log collection of each managed control plane.
```

* As an administrator adopting vector, I want to deploy it separately from my existing fluentd deployment so they can operate side-by-side and I can migrate my workloads

> 
> 
> 

Reviewer: ```suggestion

```

### Goals

* Cluster administrators control which users are allowed to define log collection and which logs they are allowed to collect.
* Users are able to continue using cluster logging and log forwarding as they have historically in the past
* Users with allowable permissions are able specify additional log collection configurations
* Log forwarder deployments are isolated so they do not interfere with other log forwarder deployments

### Non-Goals

* Introduction of the next version of logging APIs.

## Proposal

> This is where we get down to the nitty gritty of what the proposal
> actually is. Describe clearly what will be changed, including all of
> the components that need to be modified and how they will be
> different. Include the reason for each choice in the design and
> implementation that is proposed here, and expand on reasons for not
> choosing alternatives in the Alternatives section at the end of the
> document.

Reviewer: ```suggestion
```

> This proposal identifies two separate workflows in order to support the legacy deployment and allow additional deployments to meet the enhancement goals.  The legacy deployment will be familiar to users of ClusterLogForwarder prior to the implimentation of this enhancement.  They should see no differences in the manor by which they use log forwarding.  The new workflow will require additional permissions to create new ClusterLogForwarders in order to limit the number of deployments for resource concerns.  Cluster administrators will need to explicitly allow additional deployments

Reviewer: ```suggestion
This proposal identifies two separate workflows in order to support the legacy deployment and allow additional deployments to meet the enhancement goals.  The legacy deployment will be familiar to users of ClusterLogForwarder prior to the implementation of this enhancement.  They should see no differences in the manner by which they use log forwarding.  The new workflow will require additional permissions to create new ClusterLogForwarders in order to limit the number of deployments for resource concerns.  Cluster administrators will need to explicitly allow additional deployments
```

The workflows make the following assumptions:

* The **cluster-logging-operator** is deployed to the *openshift-logging* namespace
* The **cluster-logging-operator** is able to watch any namespace

> ### ClusterLogForwarder not NS:openshift-logging/Name:instance

Reviewer: The title is very confusing for the non-crew members I believe. WDYT of:

```suggestion
### Multiple-Instance Mode: Allowing multiple ClusterForwarder and ClusterLogging resources
```

**cluster administrator** is a user:

* responsible for maintaining the cluster
* able to bind cluster roles to serviceaccounts  
* that deploys the **cluster-logging-operator**

**namespace administrator** is a user:

> * able to create a serviceaccount

Reviewer: Only a serviceaccount? Not additionally secrets of type serviceaccount-token?

Author: @periklis this may be where I lack following changes in OpenShift and service accounts.  Remind me?  OpenShift will no longer be minting SA tokens?  I don't know it matters from the perspective of this enhancement other then needing a token to potentially push to a "default" logstore. We will ask the user to provide the permissions they require for CLF but the operator will need to apply any bindings required to do collection (e.g. mount the node FS for the collector)

Reviewer: Yes since 4.12 the k8s core will not create the SA token in a separate secret automatically. You need to create one Secret yourself. As you say it is only needed for potential authentication in-cluster.

Author: > Yes since 4.12 the k8s core will not create the SA token in a separate secret automatically. You need to create one Secret yourself. As you say it is only needed for potential authentication in-cluster.

What is it we need to express then here?  Are there usecases to creating an SA without a token?  Are we that usecase?

Reviewer: When we use typically ClusterLogForwarder outputs to external systems (i.e. Splunk, Elasticsearch, etc.) we ask users to give us their credentials. In this case we do not need to have a serviceacount token because the systems leave outside the cluster.

However ClusterLogForwarder outputs can point also to system living inside the same cluster w/o our product managing them (i.e. custom Loki, custom ES8). In this case the users might want to the serviceaccount token to authenticate our collector with their storage solution.

```yaml
    apiVersion: "logging.openshift.io/v1"
    kind: ClusterLogForwarder
    metadata:
      name: audit-collector
      namespace: openshift-kube-apiserver
    spec:
      pipelines:
       - inputRefs: 
         - audit
         outputRefs:
         - loki
      outputs:
      - name: loki
        type: loki
        url: https://mycenteralizedserver.some.place
```

* The namespace administrator creates a **ClusterLogging** CR with a name that matches the service account that only includes a spec for collection

```yaml
    apiVersion: "logging.openshift.io/v1"
    kind: "ClusterLogging"
    metadata:
      name: audit-collector
      namespace: openshift-kube-apiserver
    spec:
      collection:
        type: "vector"
```

##### Verification and Validations 
The operator willvalidate resources upon reconciliation of a **ClusterLogForwarder** and **ClusterLogging** CR.  Failure to meet any of the following conditions will stop the operator from deploying a collector and it will add error status to the resource.

* The **ClusterLogForwarder** CR defines a valid spec
* There exists a serviceaccount that is named the same as the **ClusterLogForwarder** CR

> * The serviceaccount is bound to clusterroles that allow the input spec of the **ClusterLogForwarder** CR

Reviewer: This is tricky to grasp, can you elaborate a bit more on:
- Are the inputs some sort of virtual resources that can be listed in a ClusterRole?
- Is the serviceaccount w/o a binding to the ClusterRole not able to send logs? Or is the ClusterLogForwarder CR in some sort of degraded state? If latter, will the cluster-logging-operator do some sort of SAR to explictly check if the serviceacount has the ClusterRole?
- In addition does it need to be a ClusterRole? AFAIU the ClusterLogForwarder resource behaves in this case namespaced.

Reviewer: > > Are the inputs some sort of virtual resources that can be listed in a ClusterRole?
> 
> The intention is for the operator manifest to include roles that map to our opinionated inputs. There will have to be documented listing of which roles are required in order to collect which input types. My expectation is we will define a convention tied to the name (e.g. collect-application-logs, collect-infrastructure-logs)
> 
> ...
> 
> Correct. The operator will perform a SAR for each source type that needs to be collected.

Both things should be documented here, the opinionated ClusterRoles for each input type as well as how the SAR looks against them. Effectively they are the core of this proposal regardless lifting naming/namespace limitations.

>  The CLF will be rejected outright if it asks for something not permitted.  We made the decision to no longer support a "degraded" state in order to be more clear regarding the actual state of collection and forwardering.
> 

How is this going to be achieved? ValidationWebhook? If yes this makes the webhook stateful because you need to talk to the API (i.e. doing SARs) and then accept or reject. Usually validation webhook are stateless and proof-check things that should never be valid inputs in the custom resource and cannot be expressed in the CRD's OpenAPI spec (e.g. two dependent fields).

Author: > Both things should be documented here, the opinionated ClusterRoles for each input type as well as how the SAR looks against them. Effectively they are the core of this proposal regardless lifting naming/namespace limitations.

I have pushed changed to make this more obvious but it was my impression this document does not include the explicit details that you are describing.  It looks more like this document is for getting enough of a gist to write JIRA cards and provide an implementation.  We have these details in the initial gdoc 

> How is this going to be achieved? ValidationWebhook? If yes this makes the webhook stateful because you need to talk to the API (i.e. doing SARs) and then accept or reject. Usually validation webhook are stateless and proof-check things that should never be valid inputs in the custom resource and cannot be expressed in the CRD's OpenAPI spec (e.g. two dependent fields).

We will not be using webhooks as we have been advised against their use but instead will make use of the existing validation model that lives in the operator and reports status.

* There exists a **ClusterLogging*** CR that is named the same as the **ClusterLogForwarder** CR
* The **ClusterLogging** CR only defines a valid collection spec.

> #### Legacy ClusterLogForwarding with ClusterLogging (NS:openshift-logging/Name:instance)

Reviewer: ```suggestion
### Legacy Mode: Allow only a single ClusterForwarder and ClusterLogging resource in openshift-logging
```

The workflows make the following assumptions:

* The **cluster-logging-operator** is deployed to the *openshift-logging* namespace
* The **cluster-logging-operator** is able to watch any namespace

### Multiple-Instance Mode: Allowing multiple ClusterForwarder and ClusterLogging resources 

This workflow supports any ClusterLogForwarder not created in the openshift-logging namespace and not named "instance".  The resource with this name in this namespace is significant to supporting the legacy workflow.

**cluster administrator** is a user:

* responsible for maintaining the cluster
* able to bind cluster roles to serviceaccounts  
* that deploys the **cluster-logging-operator**

**namespace administrator** is a user:

* able to create a serviceaccount
* manages a **ClusterLogForwarder** custome resource

> The general worklow:

Reviewer 2: ```suggestion
The general workflow:
```

```yaml
    apiVersion: "logging.openshift.io/v1"
    kind: ClusterLogForwarder
    metadata:
      name: audit-collector
      namespace: openshift-kube-apiserver
    spec:
      pipelines:
       - inputRefs:
         - audit
         outputRefs:
         - loki
      outputs:
      - name: loki
        type: loki
        url: https://mycenteralizedserver.some.place
```

* The namespace administrator creates a **ClusterLogging** CR with a name that matches the service account that only includes a spec for collection

```yaml
    apiVersion: "logging.openshift.io/v1"
    kind: "ClusterLogging"
    metadata:
      name: audit-collector
      namespace: openshift-kube-apiserver
    spec:
      collection:
        type: "vector"
```

##### Verification and Validations 
The operator will validate resources upon reconciliation of a **ClusterLogForwarder** and **ClusterLogging** CR.  Failure to meet any of the following conditions will stop the operator from deploying a collector and it will add error status to the resource.

* The **ClusterLogForwarder** CR defines a valid spec

> * There exists a serviceaccount that is named the same as the **ClusterLogForwarder** CR

Reviewer 2: I like the same-name convention for associating our own resources (CLF, CL, future additional resources) but specifically for service accounts: perhaps we should follow the same spec.serviceAccount convention that is used for Pods, since this is already familiar to users. I'm not dead set either way, but seems like there's a case for being consistent with pre-existing k8s convention here.

Author: @alanconway My original suggestion was for the spec to tell us the SA to use and we went away from that for reasons I do not recall.  I think by moving to this pattern it will also resolve my other concerns regarding how we keep the "legacy" workflow from being a unicorn from the mutli-cluster workflow.  By specifying the SA, we should be able to migrate the legacy to new naming patterns and only need to require or rely upon only a specially named SA to support things like cloudwatch upgrade.  What say you?.... `spec.serviceAccount` or name convention?

Reviewer 3: Prefer specifying the SA name instead of a magic value.

* The serviceaccount is bound to clusterroles that allow the input spec of the **ClusterLogForwarder** CR
* There exists a **ClusterLogging*** CR that is named the same as the **ClusterLogForwarder** CR
* The **ClusterLogging** CR only defines a valid collection spec.

The previous example identifies a valid **ClusterLogForwarder** CR that specs audit logs that are forwarded to a loki stack.  The following example would be rejected by the operator because it specs collection of application logs but does not have the required role binding:

```yaml
    apiVersion: "logging.openshift.io/v1"
    kind: ClusterLogForwarder
    metadata:
      name: audit-collector
      namespace: openshift-kube-apiserver
    spec:
      pipelines:
       - inputRefs:
         - audit
         - application
         outputRefs:
         - loki
      outputs:
      - name: loki
        type: loki
        url: https://mycenteralizedserver.some.place
```

> 

Reviewer 2: Suggest an additional section

#### Use of ClusterLogging

In multiple instance mode the ClusterLogging resource is _optional_. This is the reverse of legacy mode where the CL is required and the CLF is optional. In multiple-instance mode, if a CL exists with the same namespace and name  as the CLF, any relevant entries from the CL will be applied to the behavior of the CLF. Irrelevant entries are ignored.
TODO: list exactly what's relevant/irrelevant, something like:
Relevant:
- ManagementState
- LogStore (defines "default" output for this CLF instance)
- Collection (defines collector type etc. for this CLF instance)
Ignored: Visualization, Curation, Forwarder

The goal here is to allow a smooth transition when we drop the CL in v2, but avoid inventing temporary APIs or alternate ways of configuring the things that are currently configured on the CL which affect the collector.

**NOTE:** Vector is the only supported collector implimentation in this mode.

**cluster administrator** is a user:

* responsible for maintaining the cluster
* able to bind cluster roles to serviceaccounts  
* that deploys the **cluster-logging-operator**

**namespace administrator** is a user:

* able to create a serviceaccount
* able to create a serviceaccount token
* manages a **ClusterLogForwarder** custome resource

The general workflow:

* The namespace administrator creates a service account to be used by a log collector.  The service account must additionally include a token if there is intent to write to log storage managed by the Red Hat OpenShift logging subsystem.
* The cluster administrator binds cluster roles to the service account for the log types they are allowed to collect (e.g. audit, infrastructure).  Several roles are added to the operator manifest and look something like:

```
    apiVersion: rbac.authorization.k8s.io/v1
    kind: ClusterRole
    metadata:
      name: collect-audit-logs
    rules:
    - apiGroups:
      - "logging.openshift.io"
      resources:
      - audit-logs
      verbs:

>       - collect

Author: These are all "virtual" resources and verbs.  The mechanism that uses roles and bindings does not care that a resource exists or a verb is codified someplace.  It gathers these permissions and simply evaluates to determine if the conditions are met without matching them against any actual resource.

tracking-link:
- https://issues.redhat.com/browse/LOG-1344
see-also:
-
replaces:
-
---

# Multi ClusterLogForwarder
## Summary

Log forwarding is functionally a "cluster singleton" where the operator explicitly only reconcilies a **ClusterLogForwarder** in the  namespace *openshift-logging* named *instance*.  This enhancement removes that restriction to allow administrators to define multiple instance of **ClusterLogForwarder** while retaining the legacy behavior.

## Motivation

### User Stories

* As an administrator of a Red Hat managed cluster, I want to RBAC my log forwarder configuration from customer admins so they can take ownership of their log forwarder needs without being able to modify mine.
* As an administrator of Hosted Control Planes, I want to deploy individual log forwarders to isolate audit log collection of each managed control plane.
* As an administrator adopting vector, I want to deploy it separately from my existing fluentd deployment so they can operate side-by-side and I can migrate my workloads.

### Goals

* Cluster administrators control which users are allowed to define log collection and which logs they are allowed to collect.

> * Users with allowable permissions are able specify additional log collection configurations

Reviewer 4: ```suggestion
* Users with allowable permissions are able to specify additional log collection configurations
```

## Proposal

### Workflow Description
This proposal identifies two separate workflows in order to support the legacy deployment and allow additional deployments to meet the enhancement goals.  The legacy deployment will be familiar to users of ClusterLogForwarder 
prior to the implementation of this enhancement.  They should see no differences in the manner by which they use log forwarding.  The new workflow will require additional permissions to create new ClusterLogForwarders in order 
to limit the number of deployments for resource concerns.  Cluster administrators will need to explicitly allow additional deployments.

The workflows make the following assumptions:

* The **cluster-logging-operator** is deployed to the *openshift-logging* namespace
* The **cluster-logging-operator** is able to watch any namespace

#### Multiple-Instance Mode: Allowing multiple ClusterForwarder and ClusterLogging resources 

This workflow supports any ClusterLogForwarder except one named "instance" in the *openshift-logging namespace*.  The resource openshift-logging/instance is significant to supporting the legacy workflow.

> **NOTE:** Vector is the only supported collector implimentation in this mode.

Reviewer 4: ```suggestion
**NOTE:** Vector is the only supported collector implementation in this mode.
```

    apiVersion: rbac.authorization.k8s.io/v1
    kind: ClusterRole
    metadata:
      name: collect-audit-logs
    rules:
    - apiGroups:
      - "logging.openshift.io"
      resources:
      - logs
      resourceNames:
      - audit
      verbs:
      - collect

```
This role allows collection of application logs and requires the namespace administor to bind the service account to the role like:

```text
    oc create clusterrolebinding kube-audit-log-collection --clusterrole=collect-audit-logs --serviceaccount=openshift-kube-apiserver:audit-collector-sa
```

* The namespace administrator creates a **ClusterLogForwarder** CR that references the serviceaccount and the inputs for which that serviceaccount is allowed to collect

```yaml
    apiVersion: "logging.openshift.io/v1"
    kind: ClusterLogForwarder
    metadata:
      name: audit-collector
      namespace: openshift-kube-apiserver
    spec:
      serviceAccount: audit-collector-sa
      pipelines:
       - inputRefs:
         - audit
         outputRefs:
         - loki
      outputs:
      - name: loki
        type: loki
        url: https://mycenteralizedserver.some.place
```

##### Use of ClusterLogging resource

> This resource is optional in multiple instance mode and is a departure from the legacy mode where a ClusterLogging resource is always required with a ClusterLogForwarder. A namespace administrator must define a **ClusterLogging** CR named the same as the **ClusterLogForwarder** CR and in the same namespace when needing to spec collector resources or placement.

Reviewer 4: One small improvement I could see: allow for an optional spec field to allow ClusterLogging CR to reference a specifically named ClusterLogForwarding CR to not always rely on them being the same name

Author: We talked about this but @alanconway preferred to rely upon naming convention instead
